src/BatchResult.php

```diff
@@ -86,19 +54,61 @@ public function hasResponseFor(RequestInterface $request)
      * @param ResponseInterface $response
      *
      * @return BatchResult
+     */
+    public function addResponse(RequestInterface $request, ResponseInterface $response);
+
+    /**
+     * Checks if a request is successful
+     *
+     * @param RequestInterface $request
+     *
+     * @return boolean
+     */
+    public function isSuccessful(RequestInterface $request);
+
+    /**
+     * Checks if a request is failed
+     *
+     * @param RequestInterface $request
+     *
+     * @return boolean
+     */
+    public function isFailed(RequestInterface $request);
+
+    /**
+     * Returns all exceptions
      *
-     * @internal
+     * @return Exception[]
      */
-    public function addResponse(RequestInterface $request, ResponseInterface $response)
-    {
-        $new = clone $this;
-        $new->responses->attach($request, $response);
+    public function getExceptions();
 
-        return $new;
-    }
+    /**
+     * Returns an exception for a request
+     *
+     * @param RequestInterface $request
+     *
+     * @return Exception
+     *
+     * @throws \UnexpectedValueException If request is not found
+     */
+    public function getExceptionFor(RequestInterface $request);
 
-    public function __clone()
-    {
-        $this->responses = clone $this->responses;
-    }
+    /**
+     * Checks if there is an exception for a request
+     *
+     * @param RequestInterface $request
+     *
+     * @return boolean
+     */
+    public function hasExceptionFor(RequestInterface $request);
+
+    /**
+     * Adds an exception
```

Comment:
Its not perfectly clear that the BatchResponse is immutable. Do the PSR interfaces explain that? we should follow the pattern of request and response in this.